File: server-mh/src/Match/Handler/MatchPhases/TurnPhase.ts
import { Phase, PlayerColor } from "../Enums";
import { MatchContext } from "../Models/MatchContex";
import { PhaseBase } from "./PhaseBase";
import { Player } from "../Models/Player";
import { TurnState } from "../Models/TurnState";
export class TurnPhase extends PhaseBase {

    public override Start(context: MatchContext): void {
        if (!context.state.players[context.state.turnState.currentPlayer].playerState.spawnedBefore &&
            context.state.turnState.repeat <= 2) {

            context.state.turnState.anotherChance = true;


        }
        if(context.state.diceState.diceValue == 6)
            context.state.turnState.hasReward = true;

    }
    public override Update(context: MatchContext): void {
        const turnState: TurnState = context.state.turnState;
        do {
            if (turnState.anotherChance) {
                turnState.anotherChance = false;
                turnState.repeat++;
            }
            else if (turnState.hasReward) {
                turnState.hasReward = false;
            }
            else if (turnState.hasOffer) {
                turnState.hasOffer = false;
            }
            else {
                turnState.currentPlayer = this.GoToNextPlayer(turnState.currentPlayer);
                turnState.repeat = 0;

            }
        } while (context.state.players[turnState.currentPlayer].playerState.isFinished &&
            context.state.winnerList.length < 3);

        if (!(context.state.players[turnState.currentPlayer].playerState.lights > 0)) {
            this.FirePlayer(context.state.players, turnState.currentPlayer);
            context.state.label.presentPlayerCount--;
            if(context.state.label.presentPlayerCount == 0)
            {
                context.state.matchEnd = true;
                context.logger.info("TurnPhase: All players are fired, match ended");
            }

        }
        context.logger.info(`TurnPhase: currentPlayer: ${turnState.currentPlayer}, repeat: ${turnState.repeat}, anotherChance: ${turnState.anotherChance}, hasReward: ${turnState.hasReward}, hasOffer: ${turnState.hasOffer}`);
        context.broadcaster.TurnStarted(turnState.currentPlayer);
        context.state.pendingPhase = Phase.Dice;


    }

    private GoToNextPlayer(playerColor: PlayerColor): PlayerColor {

        const next = (playerColor + 1) % 4;

        return next as PlayerColor;
    }
    private FirePlayer(players: Player[], playerColor: PlayerColor): void {


        players[playerColor].playerState.isPresent = false;
        players[playerColor].playerState.isBot = true;
        players[playerColor].presence = null;
        players[playerColor].playerState.lights = 0;
        
    }



}

File: Client-unity/Ludo/Assets/Scripts/OfflineMatchHandler/TurnPhase.cs
namespace Ludo.Offline
{
    public class TurnPhase : PhaseBase
    {
        public override void Start(MatchContext context)
        {
            var turnState = context.State.TurnState;

            if (turnState.CurrentPlayer == null)
            {
                turnState.CurrentPlayer = PlayerColor.Blue;
                return;
            }

            var player =
                context.State.Players[(int)turnState.CurrentPlayer.Value];

            if (!player.PlayerState.SpawnedBefore &&
                turnState.Repeat < 2)
            {
                turnState.AnotherChance = true;
            }

            if (context.State.DiceState.DiceValue == 6)
            {
                turnState.HasReward = true;
            }
        }

        public override void Update(MatchContext context)
        {
            var state = context.State;
            var turnState = state.TurnState;

            do
            {
                if (turnState.AnotherChance)
                {
                    turnState.AnotherChance = false;
                    turnState.Repeat++;
                }
                else if (turnState.HasReward)
                {
                    turnState.HasReward = false;
                }
                else if (turnState.HasOffer)
                {
                    turnState.HasOffer = false;
                }
                else
                {
                    turnState.CurrentPlayer =
                        GoToNextPlayer(turnState.CurrentPlayer.Value);
                    GameManager.Instance.GamePlayHandler.LastContext.CurrentPlayer = turnState.CurrentPlayer.Value;

                    turnState.Repeat = 0;
                }

            } while (
                state.Players[(int)turnState.CurrentPlayer.Value]
                    .PlayerState.IsFinished
                &&
                state.WinnerList.Length < 3
            );

            var currentPlayer =
                state.Players[(int)turnState.CurrentPlayer.Value];

            if (currentPlayer.PlayerState.Lights <= 0)
            {
                state.Label.PresentPlayerCount--;

                if (state.Label.PresentPlayerCount == 0)
                {
                    // state.MatchEnd = true;
                }
            }

            context.CommandHandler.Enqueue(
    new TurnStartedCommand(turnState.CurrentPlayer.Value)
);

            state.PendingPhase = Phase.Dice;
        }

        private PlayerColor GoToNextPlayer(PlayerColor playerColor)
        {
            return (PlayerColor)(((int)playerColor + 1) % 4);
        }

        private void FirePlayer(
            Player[] players,
            PlayerColor playerColor)
        {
            var player = players[(int)playerColor];

            player.PlayerState.IsBot = true;
            player.PlayerState.Lights = 0;
        }
    }
}